go/vt/vtadmin/cluster/discovery/discovery_consul.go

```diff
@@ -184,28 +211,32 @@ func (c *ConsulDiscovery) DiscoverVTGateAddr(ctx context.Context, tags []string)
 	span, ctx := trace.NewSpan(ctx, "ConsulDiscovery.DiscoverVTGateAddr")
 	defer span.Finish()
 
-	vtgate, err := c.discoverVTGate(ctx, tags)
+	executeFQDNTemplate := false
+
+	vtgate, err := c.discoverVTGate(ctx, tags, executeFQDNTemplate)
 	if err != nil {
 		return "", err
 	}
 
-	buf := bytes.NewBuffer(nil)
-	if err := c.vtgateAddrTmpl.Execute(buf, vtgate); err != nil {
+	addr, err := textutil.ExecuteTemplate(c.vtgateAddrTmpl, vtgate)
+	if err != nil {
 		return "", fmt.Errorf("failed to execute vtgate address template for %v: %w", vtgate, err)
 	}
 
-	return buf.String(), nil
+	return addr, nil
 }
 
 // DiscoverVTGates is part of the Discovery interface.
 func (c *ConsulDiscovery) DiscoverVTGates(ctx context.Context, tags []string) ([]*vtadminpb.VTGate, error) {
 	span, ctx := trace.NewSpan(ctx, "ConsulDiscovery.DiscoverVTGates")
 	defer span.Finish()
 
-	return c.discoverVTGates(ctx, tags)
+	executeFQDNTemplate := true
+
+	return c.discoverVTGates(ctx, tags, executeFQDNTemplate)
 }
 
-func (c *ConsulDiscovery) discoverVTGates(_ context.Context, tags []string) ([]*vtadminpb.VTGate, error) {
+func (c *ConsulDiscovery) discoverVTGates(_ context.Context, tags []string, executeFQDNTemplate bool) ([]*vtadminpb.VTGate, error) {
```

Comment:
Despite looking all at the callsites, I'm not sure I follow why the `executeFQDNTemplate` boolean is necessary, or when I'd want to use true vs. false. 🤔 Maybe a comment would help? (Same comment about `discoverVtctld`'s parameter.)

Reply:
In practice it's an optimization for `Discover{Vtctld,VTGate}Addr`. If we didn't have this flag to thread through, then we would make one `ExecuteTemplate` call per gate (read also "vtctld") as part of the discovery (to generate the FQDN), pick one gate, and make a final `ExecuteTemplate` call to generate the dialable address, for a total of N+1 calls to `ExecuteTemplate`.

Since everything that's available to `ExecuteTemplate` when we're generating the address is also available when generating the FQDN, we can force the user to (potentially*) be a little redundant in the config and save the N calls for generating FQDNs and make a single `ExecuteTemplate` call for the final address.

\* "potentially" only because it's redundant if your dial address is something like "FQDN + gRPC port", then you would need a config of:

```
-fqdn-tmpl "{{ .Hostname }}.example.com"
-addr-tmpl "{{ .Hostname }}.example.com:15000"
```

whereas if we didn't do this optimization then you could specify something like

```
-fqdn-tmpl "{{ .Hostname }}.example.com"
-addr-tmpl "{{ .FQDN }}:15000"
```